### ISSUE TITLE
Parsing metadata with ComplexType using TypeDefinition is not possible

### ISSUE THREAD

**Issue by heimwege** [odata2ts/odata2ts#177]
Parsing a metadata document containing a complex type that uses a type definition results in the error message `Couldn't determine model type for data type with name 'MyService.myProperty'`.

Sample:

```
[...]
<ComplexType Name="MyComplexType">
    <Property Name="myProperty" Type="MyService.myProperty"/>
</ComplexType>
[...]
<TypeDefinition Name="myProperty" UnderlyingType="Edm.String"/>
[...]
```

Manually adjusting the metadata and skipping the [TypeDefinition](http://docs.oasis-open.org/odata/odata/v4.0/csprd02/part3-csdl/odata-v4.0-csprd02-part3-csdl.html#_Toc360208825) tag works:
```
[...]
<ComplexType Name="MyComplexType">
    <Property Name="myProperty" Type="Edm.String"/>
</ComplexType>
[...]
```

**Comment by texttechne** [comment]
Hi @heimwege,

TypeDefinitions are not supported currently. This is not restricted to ComplexTypes.

Shouldn't be that hard to implement... I'll try to take a look.

**Comment by heimwege** [comment]
The complex type bundles messages from [SAP Event Mesh](https://help.sap.com/docs/SAP_EM/bf82e6b26456494cbdd197057c09979f/df532e8735eb4322b00bfc7e42f84e8d.html) and the type definitions contain the allowed values of the respective property (the metadata itself is generated out of the CDS notation of the [SAP Cloud Application Programming Model](https://cap.cloud.sap/docs/)).

```
<ComplexType Name="messageResponse">
    <Property Name="responseCode" Type="MyService.responseCode"/>
    <Property Name="severity" Type="MyService.severity"/>
    [...]
</ComplexType>
[...]
<TypeDefinition Name="responseCode" UnderlyingType="Edm.String"/>
<TypeDefinition Name="severity" UnderlyingType="Edm.String"/>
[...]
<Annotations Target="MyService.responseCode">
     <Annotation Term="Validation.AllowedValues">
         <Collection>
             <Record Type="Validation.AllowedValue">
                 <Annotation Term="Core.SymbolicName" String="ENTITY_NOT_FOUND"/>
                 <PropertyValue Property="Value" String="ENTITY_NOT_FOUND"/>
            </Record>
            [...]
        </Collection>
    </Annotation>
</Annotations>
<Annotations Target="MyService.severity">
    <Annotation Term="Validation.AllowedValues">
        <Collection>
            <Record Type="Validation.AllowedValue">
                <Annotation Term="Core.SymbolicName" String="success"/>
                <PropertyValue Property="Value" String="SUCCESS"/>
            </Record>
            [...]
        </Collection>
    </Annotation>
</Annotations>
[...]
```

Legend:
- [Validation Vocabulary](https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Validation.V1.xml)

**Comment by texttechne** [comment]
Uhhh new address and improved design for the CAP docs. Nice.

Thanks for the example and the explanation!

Just to be sure: Annotations are currently also not supported. It's on my roadmap and captured in this issue: #149 
So I would only implement the `TypeDefinition` recognitio, so that it maps to the correct primitive type. Would this be enough for now?

**Comment by heimwege** [comment]
Currently I need to manually adjust the metadata to not get an exception during the odata2ts generation. So the mapping is exactly what I need and should be absolutely sufficient 🐱

**Comment by texttechne** [comment]
Hi @heimwege,

should be fixed with version `0.25.0` of `@odata2ts/odata2ts`.
Please try it out and close the issue if it has been fixed.

**Comment by heimwege** [comment]
fixed 👍
